### PR TITLE
vim.fn: throw error when trying to use API function

### PR DESCRIPTION
**Pull request by smolck**

### CODE REVIEW

File: src/nvim/lua/vim.lua [hunk 1]
```diff
@@ -253,9 +253,17 @@ end
 -- vim.fn.{func}(...)
 vim.fn = setmetatable({}, {
   __index = function(t, key)
-    local function _fn(...)
-      return vim.call(key, ...)
+    local _fn
+    if vim.api[key] ~= nil then
+      _fn = function(...)
+        error(string.format("Tried to call API function with vim.fn: use vim.api.%s() instead", key))
```
Comment (bfredl) [inline src/nvim/lua/vim.lua:259]: can drop the `()` (seems to suggest no args)

**Comment by bfredl** [comment]
cool, maybe a test? :)

File: test/functional/lua/vim_spec.lua [hunk 2]
```diff
@@ -715,6 +715,11 @@ describe('lua stdlib', function()
     eq({false, 'Vim:E714: List required'}, exec_lua([[return {pcall(vim.fn.add, "aa", "bb")}]]))
   end)
 
+  it('vim.fn should error when calling API function', function()
+      eq({false, 'vim.lua:259: Tried to call API function with vim.fn: use vim.api.nvim_get_current_line instead'},
```
Comment (bfredl) [inline test/functional/lua/vim_spec.lua:719]: hmm probably better to use `pcall_err` to get rid of the line number (too noisy with changes)

**Comment by kassio** [comment]
Sharing just because more people might face the same problem I'm having.

This commit broke quite a few plugins I use. Based on the error: `E5108: Error executing lua vim.lua:269: Tried to call API function with vim.fn: use vim.api.nvim_create_buf instead` I found bugs on (not exhaustive list)

- plenary.nvim
- nvim-lspconfig
- popup.nvim

Given that plenary and popup are base for other plugins I imagine a bigger impact. 

<details>
<summary>grep command</summary>

```console
$ rg 'fn.nvim_create_buf'
site/pack/packer/start/telescope.nvim/lua/telescope/builtin/internal.lua
320:        -- local buf = vim.fn.nvim_create_buf(false, true)

site/pack/packer/start/plenary.nvim/scratch/window_test.lua
8:  local bufnr = vim.fn.nvim_create_buf(false, true)
20:  buf_text = vim.fn.nvim_create_buf(false, true)
22:  buf_border = vim.fn.nvim_create_buf(false, true)

site/pack/packer/start/plenary.nvim/lua/plenary/popup.lua
37:    bufnr = vim.fn.nvim_create_buf(false, true)

site/pack/packer/start/plenary.nvim/lua/plenary/window/float.lua
50:  local bufnr = options.bufnr or vim.fn.nvim_create_buf(false, true)
91:  local primary_bufnr = vim.fn.nvim_create_buf(false, true)
178:  local bufnr = options.bufnr or vim.fn.nvim_create_buf(false, true)

site/pack/packer/start/nvim-lspconfig/lua/lspconfig/_lspui.lua
94:  local bufnr = options.bufnr or vim.fn.nvim_create_buf(false, true)

site/pack/packer/start/popup.nvim/lua/popup/init.lua
38:    bufnr = vim.fn.nvim_create_buf(false, true)
```

</details>

I'm not sure how to proceed, again, just want to share what I'm experiencing.

**Comment by smolck** [comment]
@kassio Since plugins had the freedom to be inconsistent in the usage of `vim.fn` versus `vim.api` (e.g. they could use `vim.fn.nvim_create_buf` _or_ `vim.api.nvim_create_buf`) and this PR enforces the usage of `vim.api` over `vim.fn` in cases where `vim.api` can be used, there will probably be multiple plugins that break for nightly/master users.

Thankfully, the fix for these errors is pretty trivial/straightforward, such that any broken plugins should be able to be fixed fairly quickly; at the very least, I was able to open PRs on the repositories you mentioned to fix the errors in those places (the one on `nvim-lspconfig` has already been merged).

**Comment by tami5** [comment]
:/ this is THE WORST PR ever. Seriously!!!! like wtf. I know it's the right move and I totally agree with the reasons given. But it broke everything! There should've been some type of announcement, coordination with other authors, or something before merge.

NO HARD FEELINGS!

**Comment by bfredl** [comment]
We did change our routines for breaking changes after this. Now there is this issue https://github.com/neovim/neovim/issues/14090 you can subscribe to for breaking changes, replacing the old wiki page (as apparently you cannot "follow" a github wiki page).

**Pull request merged**